toolchain/check/handle_impl.cpp

```diff
@@ -34,34 +38,143 @@ auto HandleImplForall(Context& context, Parse::ImplForallId parse_node)
 
 auto HandleTypeImplAs(Context& context, Parse::TypeImplAsId parse_node)
     -> bool {
-  auto self_id = context.node_stack().PopExpr();
-  context.node_stack().Push(parse_node, self_id);
+  auto [self_node, self_id] = context.node_stack().PopExprWithParseNode();
+  auto self_type_id = ExprAsType(context, self_node, self_id);
+  context.node_stack().Push(parse_node, self_type_id);
+  // TODO: `Self` should come into scope here, at least if it's not already in
+  // scope. Check the design for the latter case.
   return true;
 }
 
-auto HandleDefaultSelfImplAs(Context& /*context*/,
-                             Parse::DefaultSelfImplAsId /*parse_node*/)
-    -> bool {
-  // TODO: Determine self_id and push it onto node stack.
+// If the specified name scope corresponds to a class, returns the corresponding
+// class declaration.
+// TODO: Should this be somewhere more central?
+static auto AsClassScope(Context& context, SemIR::NameScopeId scope_id)
```

Comment:
```suggestion
static auto TryAsClassScope(Context& context, SemIR::NameScopeId scope_id)
```

Maybe TryAs to reflect the optional result?